pkg/objstore/bos/bos.go

```diff
@@ -0,0 +1,392 @@
+			return errors.Wrap(err, "failed to upload bos object")
+		}
+
+		return nil
+	}
+
+	result, err := b.client.BasicInitiateMultipartUpload(b.name, name)
+	if err != nil {
+		return errors.Wrap(err, "failed to initiate multi-part upload")
+	}
+
+	uploadEveryPart := func(partSize int64, part int, uploadId string) (string, error) {
+		body, err := bce.NewBodyFromSizedReader(r, partSize)
+		if err != nil {
+			return "", errors.Wrap(err, "failed to NewBodyFromSectionFile")
+		}
+
+		etag, err := b.client.UploadPart(b.name, name, uploadId, part, body, nil)
+		if err != nil {
+			if err := b.client.AbortMultipartUpload(b.name, name, uploadId); err != nil {
+				return etag, errors.Wrap(err, "failed to abort multi-part upload")
+			}
+			return etag, errors.Wrap(err, "failed to upload multi-part part")
+		}
+		return etag, nil
+	}
+
+	var parts []api.UploadInfoType
+
+	for part := 1; part <= partNums; part++ {
+		etag, err := uploadEveryPart(partSize, part, result.UploadId)
+		if err != nil {
+			return errors.Wrap(err, "failed to upload every part")
+		}
+		parts = append(parts, api.UploadInfoType{PartNumber: part, ETag: etag})
+	}
+
+	if lastSlice != 0 {
+		etag, err := uploadEveryPart(lastSlice, partNums+1, result.UploadId)
+		if err != nil {
+			return errors.Wrap(err, "failed to upload the last part")
+		}
+		parts = append(parts, api.UploadInfoType{PartNumber: partNums + 1, ETag: etag})
+	}
+
+	if _, err := b.client.CompleteMultipartUploadFromStruct(b.name, name, result.UploadId, &api.CompleteMultipartUploadArgs{Parts: parts}); err != nil {
+		return errors.Wrap(err, "failed to set multi-part upload completed")
+	}
+	return nil
+}
+
+// Iter calls f for each entry in the given directory (not recursive). The argument to f is the full
+// object name including the prefix of the inspected directory.
+func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error, opt ...objstore.IterOption) error {
+	if dir != "" {
+		dir = strings.TrimSuffix(dir, objstore.DirDelim) + objstore.DirDelim
+	}
+
+	delimiter := objstore.DirDelim
+
+	if objstore.ApplyIterOptions(opt...).Recursive {
+		delimiter = ""
+	}
+
+	var marker string
+	for {
+		if err := ctx.Err(); err != nil {
+			return errors.Wrap(err, "context closed while iterating bucket")
+		}
+
+		objects, err := b.client.ListObjects(b.name, &api.ListObjectsArgs{
+			Delimiter: delimiter,
+			Marker:    marker,
+			MaxKeys:   1000,
+			Prefix:    dir,
+		})
+
+		if err != nil {
+			return errors.Wrap(err, "listing baidu bos bucket failed")
+		}
+
+		marker = objects.NextMarker
+		for _, object := range objects.Contents {
+			if err := f(object.Key); err != nil {
+				return errors.Wrapf(err, "callback func invoke for object %s failed ", object.Key)
+			}
+		}
+
+		for _, object := range objects.CommonPrefixes {
+			if err := f(object.Prefix); err != nil {
+				return errors.Wrapf(err, "callback func invoke for object %s failed ", object.Prefix)
```

Comment:
Spurious blank line and the end here + please update the error messages according to the form in the previous comment.

Reply:
Thanks, I have been revised, please review again.